### PR TITLE
Fancy private

### PR DESCRIPTION
**Pull request by vchuravy**
Support for:

```
@kernel function new_private()
    @private mem = 1
    mem
end
```

### CODE REVIEW

**Comment by vchuravy** [comment]
@mwarusz this might also fix you use-case from #98

**Comment by vchuravy** [comment]
we should turn `mem = @private T dims` into `@private mem = ScratchArray{T, dims}`

**Comment by vchuravy** [comment]
bors try

File: src/macros.jl [hunk 1]
```diff
@@ -212,13 +226,34 @@ function emit(loop)
...
     stmts = Any[]
     append!(stmts, loop.allocations)
+
+    # private_allocations turn into lhs = ntuple(i->rhs, length(__workitems_iterspace()))
+    N = gensym(:N)
+    push!(stmts, :($N = length($__workitems_iterspace())))
+
+    for stmt in loop.private_allocations
+        if @capture(stmt, lhs_ = rhs_)
+            push!(stmts, :($lhs = ntuple(_->$rhs, $N)))
```
Comment (vchuravy) [inline src/macros.jl:236]: @simonbyrne you asked if we could support something like this:

```
e = @index(Group, Linear)
i, j = @index(Local, NTuple)
@private mem = Fprog[i,j,:,e]
```

We would need to scan `rhs` for any uses of `loop.indicies` and then project `i`/`_` onto `__workitems_iterspace` inside to obtain a valid `idx` and then instantiate the index expressions `$(loop.indicies...)`. Doable, but pfff.

Comment (simonbyrne) [reply src/macros.jl:236]: In second thought, I don't think this is really necessary.

**Comment by simonbyrne** [comment]
it would be useful to support loop indices in `@private`, i.e.
```
i, j = @index(Local, NTuple)
@private f = foo(i)
```

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
> it would be useful to support loop indices in `@private`, i.e.
> 
> ```
> i, j = @index(Local, NTuple)
> @private f = foo(i)
> ```

@vchuravy any chance we could support this?

**Comment by vchuravy** [comment]
I thought you said it isn't necessary https://github.com/JuliaGPU/KernelAbstractions.jl/pull/175/files#r553653540 ;)

**Comment by simonbyrne** [comment]
oh, i thought that was about something else.

Why you do the same thing you do for loops?

**Pull request merged**